service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseRoleServiceTest.java

```diff
@@ -0,0 +1,197 @@
+            Set<String> localAuthorityUsers = Set.of(USER_1_ID, USER_2_ID, USER_3_ID);
+            Set<String> excludedUser = emptySet();
+
+            when(organisationService.findUserIdsInSameOrganisation(LOCAL_AUTHORITY)).thenReturn(localAuthorityUsers);
+
+            doThrow(new RuntimeException())
+                .when(caseUser).updateCaseRolesForUser(any(), any(), any(), eq(USER_1_ID), any());
+
+            GrantCaseAccessException expectedException =
+                new GrantCaseAccessException(CASE_ID, Set.of(USER_1_ID), CASE_ROLES);
+
+            GrantCaseAccessException actualException = assertThrows(GrantCaseAccessException.class,
+                () -> caseRoleService.grantAccessToLocalAuthority(CASE_ID, LOCAL_AUTHORITY, CASE_ROLES, excludedUser));
+
+            assertThat(actualException).isEqualTo(expectedException);
+
+            verifyAccessGrantAttempted(USER_1_ID);
+            verifyAccessGrantAttempted(USER_2_ID);
+            verifyAccessGrantAttempted(USER_3_ID);
+        }
+
+        @Test
+        void shouldThrowsExceptionWhenGrantAccessPreparationFails() {
+            when(organisationService.findUserIdsInSameOrganisation(LOCAL_AUTHORITY))
+                .thenReturn(Set.of(USER_1_ID, USER_2_ID));
+
+            doThrow(new RuntimeException()).when(idam).authenticateUser(any(), any());
+
+            GrantCaseAccessException expectedException =
+                new GrantCaseAccessException(CASE_ID, Set.of(USER_1_ID, USER_2_ID), CASE_ROLES);
+
+            GrantCaseAccessException actualException = assertThrows(GrantCaseAccessException.class,
+                () -> caseRoleService.grantAccessToLocalAuthority(CASE_ID, LOCAL_AUTHORITY, CASE_ROLES, emptySet()));
+
+            assertThat(actualException).isEqualTo(expectedException);
+        }
+
+        @Test
+        void shouldThrowsExceptionWhenLocalAuthorityUsersCannotBeFetch() {
+            Exception fetchUsersException = new RuntimeException(format("Can not fetch users for %s", LOCAL_AUTHORITY));
+
+            doThrow(new RuntimeException()).when(organisationService).findUserIdsInSameOrganisation(LOCAL_AUTHORITY);
+
+            GrantCaseAccessException expectedException =
+                new GrantCaseAccessException(CASE_ID, LOCAL_AUTHORITY, CASE_ROLES, fetchUsersException);
+
+            GrantCaseAccessException actualException = assertThrows(GrantCaseAccessException.class,
+                () -> caseRoleService.grantAccessToLocalAuthority(CASE_ID, LOCAL_AUTHORITY, CASE_ROLES, emptySet()));
+
+            assertThat(actualException).isEqualTo(expectedException);
+        }
+    }
+
+    private void verifyAccessGrantAttempted(String userId) {
+        verify(caseUser).updateCaseRolesForUser(AUTH_TOKEN, SERVICE_AUTH_TOKEN, CASE_ID, userId, caseUser(userId));
+    }
+
+    private void verifyAccessGrantNotAttempted(String userId) {
+        verify(caseUser, never()).updateCaseRolesForUser(any(), any(), any(), eq(userId), any());
```

Comment:
I see method not only as way of reusing code but a way of adding logical name to what code does. From high perspective i guess it is easier to understand what test does with:

            verifyAccessGrantNotAttempted(USER_2_ID);
            verifyAccessGrantAttempted(USER_1_ID);
            verifyAccessGrantAttempted(USER_3_ID);

rather than :

            verify(caseUser, never()).updateCaseRolesForUser(any(), any(), any(), eq(userId), any());
            verifyAccessGrantAttempted(USER_1_ID);
            verifyAccessGrantAttempted(USER_3_ID);

and looks consistent and nicer :)